src/reconcile/rec_write.c

```diff
@@ -1409,17 +1409,32 @@ __rec_txn_read(WT_SESSION_IMPL *session, WT_RECONCILE *r,
 			*updp = upd;
 	}
 
+	/* Keep track of the selected update. */
+	upd = *updp;
+
 	/* Reconciliation should never see an aborted or reserved update. */
-	WT_ASSERT(session, *updp == NULL ||
-	    ((*updp)->txnid != WT_TXN_ABORTED &&
-	    (*updp)->type != WT_UPDATE_RESERVE));
+	WT_ASSERT(session, upd == NULL ||
+	    (upd->txnid != WT_TXN_ABORTED && upd->type != WT_UPDATE_RESERVE));
 
 	/* If all of the updates were aborted, quit. */
 	if (first_txn_upd == NULL) {
-		WT_ASSERT(session, *updp == NULL);
+		WT_ASSERT(session, upd == NULL);
 		return (0);
 	}
 
+	/* If no updates were skipped, record that we're making progress. */
+	if (upd == first_txn_upd)
+		r->update_used = true;
```

Comment:
I changed this back to checking the update before resolving birthmarks.  It may not be a big deal because this is only used as a heuristic, but I think of this as tracking whether reconciliation selected the most recent update in some list.

Maybe it would be better to track whether `all_visible` is ever true during reconciliation (so some update chain will go away during update/restore).  That's a real behavior change, so probably deserves a separate ticket and a round of performance tests.